File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.cpp
class Solution {
public:
    bool isValidMinForce(vector<int>& position,int m,int minForce) 
    {
        int ballCount=1;
        int ballPosition=position[0];
        for(int i=0;i<position.size();i++)
        {
            if(position[i]-ballPosition>=minForce)
            {
                ballPosition=position[i];
                ballCount++;
            }
            if(ballCount>=m)
            return true;
        }
        return false;
    }

    int maxDistance(vector<int>& position,int m) 
    {
        sort(position.begin(), position.end());
        int start=1;
        int end=position.back()-position.front();
        int result=0;

        while (start<=end) 
        {
            int mid=start+(end-start)/2;
            if (isValidMinForce(position,m,mid)) 
            {
                start=mid+1;
                result=mid;
            } 
            else 
                end=mid-1;
        }
        return result;
    }
};

File: 1675-magnetic-force-between-two-balls/magnetic-force-between-two-balls.java
class Solution {

    private boolean isValid(int[] position, int totalBalls, int gap){
        int ballsPlaced = 1, prevPosition = position[0];
        for(int i = 1; i < position.length; i++){
            if(position[i] - prevPosition >= gap){
                prevPosition = position[i];
                ballsPlaced++;
            }
            if(ballsPlaced == totalBalls) 
            return true;
        }
        return false;
    }

    public int maxDistance(int[] position, int balls)
    {
        Arrays.sort(position);
        int end = position[position.length - 1] - position[0];
        int start = 1;

        int maxForce = end;
        while(start <= end)
        {
            int mid = start + (end - start) /2;
            if(isValid(position, balls, mid))
            {
                maxForce = mid;
                start = mid + 1;
            }
            else
            end = mid - 1;
        }
        return maxForce;
    }
}